File: src/lib/repositories/events.ts
import type { Event, EventStatus } from '../types';
import type { EventRepository } from './interfaces';

type EventRow = {
  id: number;
  name: string;
  description: string | null;
  start_date_time: Date;
  end_date_time: Date;
  location: string | null;
  total_capacity: number;
  available_tickets: number;
  price: string;
  status: EventStatus;
  version: number | null;
  created_at: Date;
  updated_at: Date;
};

const COLUMNS =
  'id, name, description, start_date_time, end_date_time, location, total_capacity, available_tickets, price, status, version, created_at, updated_at';

function toEvent(row: EventRow): Event {
  return {
    id: row.id,
    name: row.name,
    description: row.description,
    startDateTime: row.start_date_time,
    endDateTime: row.end_date_time,
    location: row.location,
    totalCapacity: row.total_capacity,
    availableTickets: row.available_tickets,
    price: row.price,
    status: row.status,
    version: row.version ?? 0,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

export const eventRepository: EventRepository = {
  async findById(q, id) {
    const res = await q.query<EventRow>(`SELECT ${COLUMNS} FROM events WHERE id = $1`, [id]);
    const row = res.rows[0];
    return row ? toEvent(row) : null;
  },

  async findByIdForUpdate(q, id) {
    const res = await q.query<EventRow>(`SELECT ${COLUMNS} FROM events WHERE id = $1 FOR UPDATE`, [id]);
    const row = res.rows[0];
    return row ? toEvent(row) : null;
  },

  async findAvailable(q) {
    const res = await q.query<EventRow>(
      `SELECT ${COLUMNS} FROM events WHERE available_tickets > 0 AND status = 'PUBLISHED' ORDER BY start_date_time, id`,
    );
    return res.rows.map(toEvent);
  },

  async decrementAvailable(q, id) {
    await q.query('UPDATE events SET available_tickets = available_tickets - 1, version = COALESCE(version, 0) + 1 WHERE id = $1', [id]);
  },

  async incrementAvailable(q, id) {
    await q.query('UPDATE events SET available_tickets = available_tickets + 1, version = COALESCE(version, 0) + 1 WHERE id = $1', [id]);
  },
};

File: backend/app/repositories/events.py
"""Equivalente a repositories/events.ts."""
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models import Event


async def find_by_id(session: AsyncSession, event_id: int) -> Event | None:
    return await session.get(Event, event_id)


async def find_by_id_for_update(session: AsyncSession, event_id: int) -> Event | None:
    result = await session.execute(select(Event).where(Event.id == event_id).with_for_update())
    return result.scalar_one_or_none()


async def find_available(session: AsyncSession) -> list[Event]:
    result = await session.execute(
        select(Event)
        .where(Event.available_tickets > 0, Event.status == "PUBLISHED")
        .order_by(Event.start_date_time, Event.id)
    )
    return list(result.scalars().all())


async def decrement_available(session: AsyncSession, event_id: int) -> None:
    event = await session.get(Event, event_id)
    if event is not None:
        event.available_tickets -= 1
        event.version = (event.version or 0) + 1


async def increment_available(session: AsyncSession, event_id: int) -> None:
    event = await session.get(Event, event_id)
    if event is not None:
        event.available_tickets += 1
        event.version = (event.version or 0) + 1
